www/include/monitoring/objectDetails/common-func.php

```diff
@@ -122,48 +222,109 @@ function hidePasswordInCommand($commandName, $hostId, $serviceId)
     $query .= ")";
     $sth = $pearDB->prepare($query);
     for ($i = 0; $i < count($arrHostTplID); $i++) {
-        $sth->bindParam(':host_id' . $i, $arrHostTplID[$i], PDO::PARAM_INT);
+        $sth->bindParam(':host_id' . $i, $arrHostTplID[$i], \PDO::PARAM_INT);
     }
     $sth->execute();
 
-    $arrHostMacroPassword = array();
+    $arrHostMacrosPassword = [];
     while ($row = $sth->fetchRow()) {
-        $arrHostMacroPassword = array_merge(
-            $arrHostMacroPassword,
-            array($row['host_macro_name'])
+        $arrHostMacrosPassword = array_merge(
+            $arrHostMacrosPassword,
+            [$row['host_macro_name'] => $row['host_macro_value']]
         );
     }
 
-    $command = '';
-    $patternMacro = '';
-    $aCommandLineWithMacros = explode(' ', $commandLineWithMacros);
-    $aCommandLineExecuted = explode(' ', $commandLineExecuted);
-    $arrMacroPassword = array_merge($arrServiceMacroPassword, $arrHostMacroPassword);
-    $patternMacro = implode('|', $arrMacroPassword);
-    $patternMacro = str_replace('$', '\\$', $patternMacro);
-
-    if (count($arrMacroPassword) && preg_match('/(' . $patternMacro . ')/', $commandLineWithMacros)) {
-        if (count($aCommandLineWithMacros) == count($aCommandLineExecuted)) {
-            for ($i = 0; $i < count($aCommandLineWithMacros); $i++) {
-                if (preg_match_all('/(' . $patternMacro . ')/', $aCommandLineWithMacros[$i], $matches)) {
-                    $pattern = $aCommandLineWithMacros[$i];
-                    foreach ($matches as $match) {
-                        if ($arrMacroPassword[$match[0]]) {
-                            $pattern = preg_replace($match, $pattern);
-                        }
-                    }
-                    $command .= ' ' . preg_replace('/\$_(HOST|SERVICE)[a-zA-Z0-9_-]+\$/', '***', $pattern);
-                } else {
-                    $command .= ' ' . $aCommandLineExecuted[$i];
-                }
-            }
-            return preg_replace('/^ /', '', $command);
+    return $arrHostMacrosPassword;
+}
+
+/**
+ * Get the list of on demand macros of a service
+ *
+ * @param int $serviceId The ID of the service
+ *
+ * @return array
+ */
+function getServiceOnDemandMacros($serviceId): array
+{
+    global $pearDB;
+
+    // Get list of templates of the service
+    $arrtSvcTpl = getListTemplates($pearDB, $serviceId);
+    $arrSvcTplID = array($serviceId);
+    foreach ($arrtSvcTpl as $svc) {
+        $arrSvcTplID[] = $svc['service_id'];
+    }
+
+    // Get list of custom password type macros from service and linked templates
+    $query = "SELECT svc_macro_name, svc_macro_value "
+        . "FROM on_demand_macro_service "
+        . "WHERE is_password = 0 "
```

Comment:
shouldn't it be is_password = 1 ?

Reply:
no this is done by getServiceMacrosPassword function

Reply:
ok